### PR TITLE
refactor: code-quality review and cleanup before the next protocol wave

### PR DESCRIPTION
**Pull request by EONRaider**
## Summary

A full code-quality review and refactor pass across source, tests, and CI, done ahead of onboarding a large batch of new protocols. The codebase was already unusually mature (99.95% coverage, strict mypy, comprehensive CI) — this is a surgical pass on the genuine gaps found, not a rewrite.

**Source deduplication:**
- Added a shared `hex_str()` helper in `_base.py`, replacing 11 duplicated hex-formatting properties across 8 files
- Added `_ip_protocol_enum()`/`_ethertype_enum()` helpers beside the existing `_ip_protocol_name()`/`_ethertype_name()`, replacing 9 duplicated enum-fallback properties across 6 files — also fixed `arp.py`'s `ptype_name`, which was hand-rolling logic `ethernet.py` already centralized, and a 4th, previously undetected copy of the same pattern in `vlan.py`
- Added a new `_tlv.py` module unifying three duplicated TLV/options-parsing loops (TCP, IPv4, IPv6 extension headers) behind one generic, bounds-checked cursor walker

**Test hardening:**
- Fixed a real gap: `DNSOverTCP` was missing from `test_fuzz.py`'s `ALL_PROTOCOLS`; added a `Protocol.__subclasses__()`-walking completeness test so this can't silently recur
- Relocated `TestEnumCompleteness` out of `test_ipv6_ext.py` into its own `test_enums.py`
- Brought `conftest.py`/`strategies.py` under strict mypy (sized the full-suite gap at 639 errors across 22 files — too large for this pass, staged incrementally)
- A scoped `mutmut` mutation-testing audit against `checksum.py`, `_tlv.py`, and DNS's name-compression walk closed several genuine test gaps (an untested IGMP checksum path, weak error-identity assertions, a signed/unsigned boundary case) and documented two accepted-equivalent-mutant survivors

**CI/CD hardening:** SHA-pinned GitHub Actions + concurrency groups on `ci.yml`/`release.yml`/`fuzz.yml`, a dependency-vulnerability-scan job, a reproducible-build hash check, a nightly scoped mutation-testing workflow, CodeQL's default Python analysis, an OSSF Scorecard workflow + README badge, a `lychee` docs link-checker, Conventional-Commits PR-title linting, and Dependabot for both the `uv` and `github-actions` ecosystems.

**Docs sync:** `ARCHITECTURE.md`'s layout map and cookbook, and `docs/CLAIMS.md`'s drift-prone figures (module count, wheel size, line/statement/raise-site counts), updated to reflect the refactor per the repo's own re-measurement convention.

**Deliberately left alone**, with reasoning preserved in the commit history: the `object.__new__`/`object.__setattr__` decode-shortcut duplication (sits on the benchmark-gated hot path; the underlying finding argued against fixing it), forcing DNS's two byte-walk functions together (different termination semantics), generalizing ARP's single-consumer enum-fallback helpers, and 5 test-redundancy candidates (flagged for a human confirmation pass, not deleted).

## Test plan

- [x] `uv run ruff check`
- [x] `uv run ruff format --check`
- [x] `uv run mypy`
- [x] `uv run pytest --cov=netprotocols --cov-report=term` — 1148 passed, 99.95% coverage (gate: 98%)
- [x] `pytest tests/test_docs.py`
- [x] `pytest tests/test_workflows.py`
- [x] `scripts/benchmark.py --check --threshold 15 --compare` — -4.8% vs. baseline, within threshold
- [ ] New CI jobs (mutation nightly, dependency audit, CodeQL, Scorecard, reproducible-build check, docs-lint, PR-title-lint) verified locally for YAML validity and logic, but GitHub-native runtime behavior (permissions, OIDC scopes, third-party action expectations) can only be confirmed once this PR's own CI run completes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01R7YhSxmmB2VT8B5inbXP6z

---
_Generated by [Claude Code](https://claude.ai/code/session_01R7YhSxmmB2VT8B5inbXP6z)_

## Summary by CodeRabbit

- **New Features**
  - Added a public helper for consistent hexadecimal value formatting.
  - Improved validation and error reporting for TCP, IPv4, and IPv6 options.

- **Security & Quality**
  - Added dependency auditing, code scanning, security scoring, fuzzing, mutation testing, reproducible-build checks, and release provenance verification.
  - Added automated pull request title and documentation link checks.

- **Documentation**
  - Updated architecture, contribution, measurement, and project documentation.

- **Testing**
  - Expanded checksum, protocol, workflow, enum, DNS, and parser coverage.

### CODE REVIEW

File: docs/CLAIMS.md [hunk 1]
```diff
@@ -375,11 +375,11 @@ the prior figure here was built from `netprotocols-2.0.0`, two releases
 behind the tree every other number in this file now describes — this
 was stale, not a real regression.*
 
-88.3 KB wheel (`uv build`, `netprotocols-2.2.0-py3-none-any.whl`,
-90,390 bytes, deterministic across repeated builds) against scapy
+90.4 KB wheel (`uv build`, `netprotocols-2.2.1-py3-none-any.whl`,
+92,553 bytes, deterministic across repeated builds) against scapy
 2.7.0's 2.47 MB (2,590,982 bytes, current PyPI `bdist_wheel`) —
```
[the remark this answers is not in the document]
Comment (EONRaider) [reply docs/CLAIMS.md]: Confirmed and fixed in 485d04b — 92,553 bytes is binary (KiB/MiB), not decimal; corrected both the wheel and scapy comparison figures in this section.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/CLAIMS.md [hunk 2]
```diff
@@ -375,11 +375,11 @@ the prior figure here was built from `netprotocols-2.0.0`, two releases
...
-**28.7× smaller**, down from the ~46× recorded at v1.3.0 and the 29.6×
-this section previously quoted from the 2.0.0-era build. 6,413 lines
+**28.0× smaller**, down from the ~46× recorded at v1.3.0 and the 29.6×
+this section previously quoted from the 2.0.0-era build. 6,466 lines
```
[the remark this answers is not in the document]
Comment (EONRaider) [reply docs/CLAIMS.md]: Confirmed and fixed in 485d04b — README.md now matches CLAIMS.md's re-measured 90.4 KiB / 28.0× figures instead of the pre-refactor 88.3 KB / ~29×.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/netprotocols/_tlv.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                f"{protocol.__name__} option length must be at least "
+                f"{min_option_length}, got {length}",
+                protocol=protocol,
+                field=field,
+                offset=cursor,
+                expected=f">={min_option_length}",
+                actual=length,
+            )
+        if length_includes_header:
+            end = cursor + length
```
[the remark this answers is not in the document]
Comment (EONRaider) [reply src/netprotocols/_tlv.py:106]: Confirmed — verified this isn't reachable through today's three call sites (TCP and IPv4 both already pass `min_option_length=2`), but it's a real latent violation of the function's own documented "never hangs" guarantee, and a real DoS risk for any future caller of this shared walker that doesn't realize the constraint. Fixed in 485d04b: the >=2 floor is now structural whenever `length_includes_header=True`, combined with whatever stricter `min_option_length` a caller passes, rather than left to caller discipline. Also added a parametrized regression test (`test_length_includes_header_enforces_a_floor_of_two_regardless`) covering declared lengths 0 and 1 against `min_option_length` of `None`, `0`, and `1`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_workflows.py [hunk 4]
```diff
@@ -20,6 +29,9 @@
 WORKFLOWS = Path(__file__).resolve().parent.parent / ".github" / "workflows"
 CI = WORKFLOWS / "ci.yml"
 RELEASE = WORKFLOWS / "release.yml"
+FUZZ = WORKFLOWS / "fuzz.yml"
+DEPENDABOT = WORKFLOWS.parent / "dependabot.yml"
+ALL_WORKFLOWS = sorted(WORKFLOWS.glob("*.yml"))
```
[the remark this answers is not in the document]
Comment (EONRaider) [reply tests/test_workflows.py]: Confirmed and fixed in 485d04b, using your suggested diff.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EONRaider** [comment]
Addressing the two outside-diff-range findings from the latest review:

- **`tests/test_workflows.py:144-145` (Major)**: confirmed and this was serious — the regex required `uses:` immediately after leading whitespace, but every actual `uses:` step in this repo's workflow files is written in the standard YAML list form (`- uses: ...`), which that pattern can't match. `test_remote_actions_are_pinned_to_a_commit_sha` had been matching **zero lines in every workflow file**, passing vacuously regardless of whether anything was actually pinned. Fixed in 9a966bd: allowed an optional `- ` list marker, extracted the matching logic into a standalone `uses_ref()` with its own direct regression tests (`TestUsesRefMatching`), and manually confirmed 38 `uses:` lines are now actually inspected (0 before) and that a deliberately unpinned action now fails the check.
- **`pyproject.toml:99` (Minor)**: confirmed and fixed in the same push — corrected the rounded coverage figure to 99.95% (2051/2052).

Both verified with the full QA ladder (ruff, mypy, pytest at 99.95% coverage) before pushing.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EONRaider** [comment]
Addressing the two outside-diff-range findings from the latest review:

- **`tests/test_workflows.py:226-232` (Major)**: confirmed — the test only checked for the word "reproducible" and `>=3` occurrences of `uv build`, which a workflow could satisfy while building three times and never comparing anything. ci.yml's actual step already does a real `sha256sum` comparison with a hard failure on mismatch (verified by reading it directly); the test now also asserts that comparison logic is present, not just its surrounding vocabulary.
- **`pyproject.toml:115-118` (Major)**: confirmed, and this one was a real, embarrassing drift — pyproject.toml's own `[tool.mutmut]` comment documents positional filters to keep the nightly run scoped to DNS's name-compression walk specifically, but `mutation.yml`'s actual step ran plain `uv run --frozen mutmut run` without them. That would have silently mutated (and nightly-reported on) all of `dns.py`'s untriaged record/RDATA/dataclass code, not the narrower slice that was actually audited. Fixed by applying the same four filters the pyproject.toml comment specifies, and added `TestMutationWorkflow` so the two configs can't silently diverge again.

Both verified with the full QA ladder (ruff, mypy, pytest at 99.95% coverage) and manual YAML validation before pushing (1491466).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_workflows.py [hunk 5]
```diff
@@ -116,3 +134,229 @@ def test_every_local_reusable_reference_resolves(self) -> None:
...
+        "netprotocols.checksum.*",
+        "netprotocols._tlv.*",
+        "netprotocols.layer7.dns.x__read_name*",
+        "netprotocols.layer7.dns.x__labels*",
+    )
+
+    def test_run_applies_the_documented_dns_scope_filters(self) -> None:
+        text = MUTATION.read_text()
+        for pattern in self.EXPECTED_FILTERS:
+            assert pattern in text, (
```
[the remark this answers is not in the document]
Comment (EONRaider) [reply tests/test_workflows.py]: Confirmed and fixed in ddb3d23, using your suggested regex (verified it matches my file's exact indentation before applying). Also manually confirmed the fix actually closes the gap: constructed a synthetic case where a filter is present in a comment but absent from the real `mutmut run` command, and verified the old substring-anywhere check would have passed it while the new block-scoped check correctly fails it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EONRaider** [comment]
@coderabbitai review

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_workflows.py [hunk 6]
```diff
@@ -116,3 +134,246 @@ def test_every_local_reusable_reference_resolves(self) -> None:
...
+        """Checks the filters inside the actual `mutmut run` argument
+        block, not just anywhere in the file -- a bare substring check
+        would still pass if a filter were dropped from the real command
+        while its text lingered in a comment (this file's own comment
+        block above the step names all four, for exactly the reason
+        that error would be easy to make)."""
+        text = MUTATION.read_text()
+        match = re.search(
+            r"(?m)^          uv run --frozen mutmut run \\\n"
+            r"(?P<filters>(?:            '[^']+'\s*\\?\n?)+)",
```
[the remark this answers is not in the document]
Comment (EONRaider) [reply tests/test_workflows.py]: Confirmed and fixed in 696733e. Per this repo's new steward skill (independent implementation + independent review for CI-reported bugs), this fix was implemented by one independent sub-agent and separately verified by a second: the second agent reproduced the old/new parsing logic standalone, ran its own adversarial cases (an appended undocumented 5th filter with a valid continuation backslash — the old substring check incorrectly passed it, the new exact-set comparison correctly fails it) against the real `mutation.yml`, and confirmed no regressions before this was pushed. The fix now parses the argument lines one at a time exactly as bash would (a line without its trailing `\` ends the real command right there, so parsing stops even if more shape-valid lines follow) and compares the complete parsed filter set against `EXPECTED_FILTERS` for exact equality rather than presence.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EONRaider** [comment]
@coderabbitai review

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .claude/skills/steward/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+2. **Delegate the fix itself** to an independent sub-agent (e.g. via the Agent tool), giving it the finding and enough context to act, but without it inheriting the current session's authorship of the original code — it should approach the fix fresh, not defend or extend its own prior work.
+3. **Delegate review of that fix** to a second, separate sub-agent before pushing — distinct from the agent that implemented the fix. This agent should verify the fix actually closes the gap (ideally by constructing a case that would have slipped past the old code and confirming the new code catches it, not just that the reported symptom is gone), and check it didn't introduce the same class of bug elsewhere.
+4. Only after both steps pass does the fix get pushed and the finding marked resolved.
+
+This applies to fixes for CI-reported bugs specifically. It does not require every commit or every piece of new feature work in this repo to go through this two-step delegation — just the fix-a-bug-CI-found loop, where the risk of an author's blind spots recurring is highest.
```
[the remark this answers is not in the document]
Comment (EONRaider) [reply .claude/skills/steward/SKILL.md]: Confirmed and fixed in 534f889 — broadened the scope wording to match the rule's own earlier lines (CI, CodeRabbit/bot, and human-reviewer findings alike), rather than narrowing it to "CI-reported bugs." Fixed directly in this session since it's a wording clarification in prose with no code/logic stakes, not the kind of fix the two-step delegation rule itself targets.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_workflows.py [hunk 8]
```diff
@@ -116,3 +134,280 @@ def test_every_local_reusable_reference_resolves(self) -> None:
...
+        # Parse the quoted argument lines exactly as bash would read
+        # them: a line ending in " \" continues the command onto the
+        # next line, and a line without one is the last argument --
+        # parsing stops there even if more quoted-looking lines follow,
+        # because bash would never see them as part of this command.
+        filters: list[str] = []
+        for line in match.group("rest").splitlines():
+            argument = re.fullmatch(r"            '([^']+)'( \\)?", line)
+            if argument is None:
+                break
```
[the remark this answers is not in the document]
Comment (EONRaider) [reply tests/test_workflows.py]: Confirmed and fixed in 5e60286. Per this repo's steward skill, the fix was implemented by an independent sub-agent (not the one that wrote the code being fixed) and adversarially reviewed by a second before being pushed.

The reviewer independently re-derived every claim rather than trusting the fixer's narrative: it verified the real `mutation.yml`'s last filter line has no trailing backslash (so the fix's `continues`-flag normal-end path is genuinely exercised, not just theoretical — it proved a naive unconditional-`assert` version of your suggested diff actually breaks on the real file's trailing comment line); ran 6 hand-traced/executed scenarios (the real file, an unquoted continuation with a different word than the fixture, a 3-of-4-filters case, a valid 5th filter, and edge cases); and reconstructed the pre-fix `break`-based loop to confirm the new regression test (`test_run_rejects_an_unquoted_argument_on_a_continued_line`) actually fails against it — proving it isn't vacuous. Full QA ladder (ruff, mypy, pytest --cov at 99.95%) is green.

The fix goes slightly beyond your suggested diff: an unconditional `assert` would itself break on the real file (the last real filter has no trailing `\`, so the line right after it — a comment — is not a continuation and must not be asserted on). It extracts the parsing into `_parse_mutmut_run_filters` with a `continues: bool` tracking whether the previous line ended in a continuation backslash, asserting only when a line that *is* a genuine continuation fails to parse as a quoted filter.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EONRaider** [comment]
Re: the "Docstring Coverage" pre-merge warning (60.80% vs. an 80% threshold) — verified, not acting on it:

- Every file this PR touches passes this repo's own docstring gate with zero violations: `uv run ruff check --select D100,D101,D103` (scoped to `src/`, exempting `tests/`/`scripts/` per `pyproject.toml`'s documented, pre-existing convention) is clean across the full diff.
- The 60.80% figure is dominated by the ~36 new test methods this PR adds. Tests in this repo rely on descriptive names instead of docstrings by established convention (see `pyproject.toml`'s per-file-ignores comment, which already counted 95 undocumented test classes / 42 undocumented test functions before this PR) — that's a deliberate, documented choice, not a gap this PR introduced.
- To sanity-check that the narrower D100/D101/D103 scope isn't just hiding a real problem: enabling the stricter `D102` (public methods) repo-wide surfaces 25 pre-existing violations, all in files this PR never touches (`dhcp.py`, `dns.py`, etc.) — confirming that rule was deliberately left out of this engagement's scope rather than overlooked.

No doc changes planned in response to this specific check. Triggering a fresh review below since the automated review paused after the last two commits and hasn't seen 534f889/5e60286 yet.

@coderabbitai review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EONRaider** [comment]
Re: the outside-diff-range finding on `tests/test_workflows.py` lines 226-250 (`TestCIWorkflow.test_build_verifies_reproducibility` accepting tokens from an echo/comment/unreachable branch rather than the step's actual executed script) — confirmed and fixed in b6902df.

Per this repo's steward skill, implemented by an independent sub-agent and adversarially reviewed by a second before being pushed. The reviewer verified the fix correctly accepts the real `ci.yml` step (indentation, the `cut -d ' ' -f1` pipe, `set -euo pipefail`, all included), rejects a cosmetic stand-in that satisfies every one of the old five assertions (confirmed non-vacuous by reconstructing the pre-fix logic against the fixture), resists decoy content in adjacent steps, and fails loudly rather than silently on step-name drift.

The reviewer also surfaced one further latent gap beyond what was reported here: the mismatch-body regex captures up to the first `fi` it encounters rather than the one that actually balances the `if`, so a deliberately nested `if false; then exit 1; fi` inside the real mismatch branch would still slip past. This doesn't affect the current `ci.yml` (no such nesting exists there), and closing it fully would mean tracking `if`/`fi` nesting depth — moving from a text heuristic toward actually parsing bash. Documented as an accepted limitation in a code comment rather than chased further, matching how this repo already documents other accepted limitations (e.g. `pyproject.toml`'s mutmut equivalent-mutant survivors).

@coderabbitai review

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_workflows.py [hunk 9]
```diff
@@ -116,3 +134,449 @@ def test_every_local_reusable_reference_resolves(self) -> None:
...
+            "          uv build -o dist-repro-2\n"
+            '          echo "sha256sum check: would compare "$hash1" '
+            '!= "$hash2" and exit 1 on mismatch"\n'
+            '          echo "Wheel build is reproducible."\n'
+        )
+        script = self._reproducibility_step_script(cosmetic_step)
+        with pytest.raises(
+            AssertionError, match="sha256sum of the dist-repro-1 wheel"
+        ):
+            self._assert_diffs_two_independent_builds(script)
```
[the remark this answers is not in the document]
Comment (EONRaider) [reply tests/test_workflows.py:365]: Confirmed and fixed in 41e2a61, using your suggested fixture (verified it works exactly as proposed — reaches and fails specifically at the `exit 1` check, without adjustment).

Per this repo's steward skill, implemented by an independent sub-agent and adversarially reviewed by a second before being pushed. The reviewer proved the new test's coverage isn't coincidental by breaking each of the two previously-unexercised checks in turn and confirming the failure mode changed as expected: neutering the `exit 1` assertion flips the test to "did not raise"; breaking the `if`/`fi` regex moves the failure to the earlier "no reachable if...fi block" assertion instead.

The reviewer also noticed an `elif`-shaped variant of the already-documented nested-`if` limitation (same accepted tradeoff — `exit 1` textually inside the captured body but unreachable when the mismatch branch actually fires) and confirmed it's the same disclosed class, not a new gap from this fix. Extended the existing "known accepted limitation" comment above the mismatch regex to name it explicitly rather than opening a new round for it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EONRaider** [comment]
@coderabbitai review

---
_Generated by [Claude Code](https://claude.ai/code)_
